### PR TITLE
feat(linter): implement `no-throw-literal`

### PR DESCRIPTION
**Pull request by shulaoda**
Related to #479

### CODE REVIEW

**Review by camc314 — APPROVED** [review]

File: crates/oxc_linter/src/rules/eslint/no_throw_literal.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+fn no_throw_literal_diagnostic(span: Span, is_undef: bool) -> OxcDiagnostic {
+    let message =
+        if is_undef { "Do not throw undefined." } else { "Expected an error object to be thrown." };
+
+    OxcDiagnostic::warn("Disallow throwing literals as exceptions")
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/no_throw_literal.rs:15]: Your help message should be the warning message. Please refer to the diagnostics message part of the "adding a lint rule" guide

File: crates/oxc_linter/src/rules/eslint/no_throw_literal.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        } else if matches!(&stmt.argument, Expression::Identifier(id) if id.name == "undefined") {
+            ctx.diagnostic(no_throw_literal_diagnostic(stmt.span, true));
+        };
+    }
+}
+
+impl NoThrowLiteral {
+    fn could_be_error(expr: &Expression) -> bool {
+        match expr {
+            Expression::Identifier(_)
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/no_throw_literal.rs:90]: Can we attempt to resolve this reference and see if it's really an error object?

File: crates/oxc_linter/src/snapshots/no_throw_literal.snap [hunk 3]
```diff
@@ -0,0 +1,121 @@
+---
+source: crates/oxc_linter/src/tester.rs
+---
+  ⚠ eslint(no-throw-literal): Disallow throwing literals as exceptions
+   ╭─[no_throw_literal.tsx:1:1]
+ 1 │ throw 'error';
```
Comment (DonIsaac) [inline crates/oxc_linter/src/snapshots/no_throw_literal.snap:6]: Please make the span only cover the thrown expression

File: crates/oxc_linter/src/rules/eslint/no_throw_literal.rs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            ctx.diagnostic(no_throw_literal_diagnostic(stmt.span, false));
+        } else if matches!(&stmt.argument, Expression::Identifier(id) if id.name == "undefined") {
+            ctx.diagnostic(no_throw_literal_diagnostic(stmt.span, true));
+        };
+    }
+}
+
+impl NoThrowLiteral {
+    fn could_be_error(expr: &Expression) -> bool {
+        match expr {
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/no_throw_literal.rs:89]: Consider using `get_inner_expr`. That might not actually be the name, I'm on my phone and cannot check

File: crates/oxc_linter/src/rules/eslint/no_throw_literal.rs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        "throw 'literal' && new Error();",
+        "throw new Error() || 'literal';",
+        "throw foo ? new Error() : 'literal';",
+        "throw foo ? 'literal' : new Error();",
+        "throw tag `${foo}`;", // { "ecmaVersion": 6 },
+        "function* foo() { var index = 0; throw yield index++; }", // { "ecmaVersion": 6 },
+        "async function foo() { throw await bar; }", // { "ecmaVersion": 8 },
+        "throw obj?.foo",      // { "ecmaVersion": 2020 },
+        "throw obj?.foo()",    // { "ecmaVersion": 2020 }
+    ];
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/no_throw_literal.rs:174]: Please add some typescript examples to both pass and fail tests.

File: crates/oxc_linter/src/rules/eslint/no_throw_literal.rs [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    /// throw e;
+    ///
+    /// try {
+    ///     throw new Error("error");
+    /// } catch (e) {
+    ///     throw e;
+    /// }
+    /// ```
+    NoThrowLiteral,
+    correctness,
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/no_throw_literal.rs:70]: todo: @camc314 we could add a conditional_suggestion when a string is being thrown

File: crates/oxc_linter/src/rules/eslint/no_throw_literal.rs (hunk 1 above)
Comment (DonIsaac) [reply crates/oxc_linter/src/rules/eslint/no_throw_literal.rs:15]: https://oxc.rs/docs/contribute/linter/adding-rules.html#diagnostics

File: crates/oxc_linter/src/rules/eslint/no_throw_literal.rs (hunk 2 above)
Comment (shulaoda) [reply crates/oxc_linter/src/rules/eslint/no_throw_literal.rs:90]: This seems a bit tricky. The only case worth discussing is in assignment scenarios, where certain things are not considered `Error` objects.

File: crates/oxc_linter/src/rules/eslint/no_throw_literal.rs [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+                        .get_expression()
+                        .map_or(true, |expr| Self::could_be_error(expr))
+                        || Self::could_be_error(&expr.right);
+                }
+
+                false
+            }
+            Expression::SequenceExpression(expr) => {
+                if expr.expressions.len() > 0 {
+                    return Self::could_be_error(expr.expressions.last().unwrap());
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/no_throw_literal.rs:137]: Don't unwrap. Use `is_some_and`

File: crates/oxc_linter/src/rules/eslint/no_throw_literal.rs [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{GetSpan, Span};
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+fn no_throw_literal_diagnostic(span: Span, is_undef: bool) -> OxcDiagnostic {
+    let message =
+        if is_undef { "Do not throw undefined" } else { "Expected an error object to be thrown" };
+
+    OxcDiagnostic::warn(message).with_label(span)
```
Comment (camchenry) [inline crates/oxc_linter/src/rules/eslint/no_throw_literal.rs:15]: Could we add help text for this? Like:

- ```
  Wrap the value with an error object: `throw new Error(...)`
  ```
- or if `undefined`:
  ```
  throw an error object instead: `throw new Error()`
  ```

Comment (shulaoda) [reply crates/oxc_linter/src/rules/eslint/no_throw_literal.rs:15]: Currently, there are many rules that do not provide help, only help prompts when fixable.

> Wrap the value with an error object: `throw new Error(...)`

For fixable types like `string` and `template`, they automatically include help suggestions. We've already implemented fixes for some of them, and the remaining ones are largely unfixable, so this prompt might not be necessary.

> throw an error object instead: `throw new Error()`

For `undefined`, this should be a very intuitive error prompt in itself, and perhaps `help` is not necessary.


Is there a necessary requirement for adding `help`? Under what circumstances should we add it? @DonIsaac

File: crates/oxc_linter/src/rules/eslint/no_throw_literal.rs (hunk 2 above)
Comment (DonIsaac) [reply crates/oxc_linter/src/rules/eslint/no_throw_literal.rs]: I've added it myself.

File: crates/oxc_linter/src/rules/eslint/no_throw_literal.rs (hunk 8 above)
Comment (camchenry) [reply crates/oxc_linter/src/rules/eslint/no_throw_literal.rs:15]: consider this comment as non-blocking. I'm of the opinion that every warning/error should include at least _some_ indication of how to fix it. I think a static message here would be fine too:

```rust
.with_help("Throw an Error object instead: `throw new Error()`")
```

Some diagnostics don't contain help messages, I would consider that something we should fix. From our [Linter Rule principles](https://oxc.rs/docs/contribute/linter/adding-rules.html#diagnostics):

> - The `message` should be an imperative statement about what is wrong, not a description of what the rule does.
> - The `help` message should be a command-like statement that tells the user how to fix the issue.

**Pull request merged**